receiver/m365receiver/logs.go

```diff
@@ -0,0 +1,514 @@
+			{"general", "Audit.General", cfg.Logs.GeneralLogs},
+			{"exchange", "Audit.Exchange", cfg.Logs.ExchangeLogs},
+			{"sharepoint", "Audit.SharePoint", cfg.Logs.SharepointLogs},
+			{"azureAD", "Audit.AzureActiveDirectory", cfg.Logs.AzureADLogs},
+			{"dlp", "DLP.All", cfg.Logs.DLPLogs},
+		},
+		root:      fmt.Sprintf("https://manage.office.com/api/v1.0/%s/activity/feed/subscriptions/content?contentType=", cfg.TenantID),
+		startRoot: fmt.Sprintf("https://manage.office.com/api/v1.0/%s/activity/feed/subscriptions/start?contentType=", cfg.TenantID),
+	}
+}
+
+// creates the client for http requests, and initializes metadata struct
+func (l *m365LogsReceiver) Start(ctx context.Context, host component.Host) error {
+	// create default client
+	httpClient, err := l.cfg.ToClient(host, l.settings)
+	if err != nil {
+		l.logger.Error("error creating HTTP client", zap.Error(err))
+		return err
+	}
+
+	// create m365 log client, create token and start audit subscriptions
+	l.client = newM365Client(httpClient, l.cfg, "https://manage.office.com/.default")
+	err = l.client.GetToken(ctx)
+	if err != nil {
+		l.logger.Error("error creating authorization token", zap.Error(err))
+		return err
+	}
+	for _, a := range l.audits {
+		err = l.client.StartSubscription(ctx, l.startRoot+a.route)
+		if err != nil {
+			l.logger.Error("error starting audit subscriptions", zap.Error(err))
+			return err
+		}
+	}
+
+	// set cancel function
+	cancelCtx, cancel := context.WithCancel(ctx)
+	l.cancel = cancel
+
+	// init checkpoint stuff
+	storageClient, err := adapter.GetStorageClient(ctx, host, l.cfg.StorageID, l.id)
+	if err != nil {
+		return fmt.Errorf("failed to get storage client: %w", err)
+	}
+	l.storageClient = storageClient
+	l.loadCheckpoint(cancelCtx)
+
+	return l.startPolling(cancelCtx)
+}
+
+func (l *m365LogsReceiver) Shutdown(ctx context.Context) error {
+	l.logger.Debug("shutting down logs receiver")
+	l.cancel()
+	l.wg.Wait()
+	return l.checkpoint(ctx)
+}
+
+// spins a go routine at each poll interval to go get logs
+func (l *m365LogsReceiver) startPolling(ctx context.Context) error {
+	t := time.NewTicker(l.pollInterval)
+	l.wg.Add(1)
+	go func() {
+		defer l.wg.Done()
+		for {
+			select {
+			case <-t.C:
+				if err := l.pollLogs(ctx); err != nil {
+					l.logger.Error("error while polling for logs", zap.Error(err))
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// spins a go routine for each audit type/endpoint
+func (l *m365LogsReceiver) pollLogs(ctx context.Context) error {
+	var st string
+	if l.record.NextStartTime != nil {
+		st = l.record.NextStartTime.Format(layout)
+	} else {
+		st = pcommon.NewTimestampFromTime(time.Now().Add(-l.pollInterval)).AsTime().UTC().Format(layout)
+	}
+
+	now := time.Now().UTC()
+
+	auditWG := &sync.WaitGroup{}
+	for i := 0; i < len(l.audits); i++ {
+		auditWG.Add(1)
+		go l.poll(ctx, now, st, &l.audits[i], auditWG)
+	}
+	auditWG.Wait()
+
+	l.record.NextStartTime = &now
```

Comment:
Since you said the logs are not always immediately available in the API, we should only update NextStartTime to be equal to the latest timestamp of log we did collect, so that we will keep requesting from the last log collected, allowing us to hopefully avoid lost data.

Reply:
Discussed offline, this shouldn't be necessary based on the behavior of the API.